cirq-core/cirq/contrib/bayesian_network/bayesian_network_gate.py

```diff
@@ -0,0 +1,198 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This class implements Quantum Bayesian Networks as described in:
+https://arxiv.org/abs/2004.14803
+
+In addition, these write ups could be helpful:
+https://towardsdatascience.com/create-a-quantum-bayesian-network-d26d7c5c4217
+https://towardsdatascience.com/how-to-create-a-quantum-bayesian-network-5b011914b03e
+
+In order to reduce the number of gates, the code uses Gray coding, as describe in separate work:
+https://arxiv.org/abs/1306.3991
+
+Note that Bayesian networks are directed acyclic graphs, but the present class does not handle any
+of the graph properties. Instead, it narrowly focuses only on the quantum implementation, and only
+receives as inputs base Python objects, not a graph. It is incumbent on the user to make sure the
+input of the gates indeed represent a Bayesian network.
+"""
+import math
+from typing import Any, cast, Dict, List, Optional, Sequence, Tuple, Union
+
+from sympy.combinatorics import GrayCode
+
+import cirq
+from cirq import value
+from cirq.ops import raw_types
+
+
+@value.value_equality
+class BayesianNetworkGate(raw_types.Gate):
+    """A gate that represents a Bayesian network."""
```

Comment:
Might want to expand this explanation here.

Reply:
Moved the docstring here, so it's quite larger now.